plugins/woocommerce-admin/client/products/hooks/use-product-entity-record.ts

```diff
@@ -0,0 +1,52 @@
+/**
+ * External dependencies
+ */
+import { AUTO_DRAFT_NAME } from '@woocommerce/product-editor';
+import { Product } from '@woocommerce/data';
+import { useDispatch, useSelect, select as WPSelect } from '@wordpress/data';
+
+import { useEffect, useState, useMemo } from '@wordpress/element';
+
+const useAutoDraft = () => {
+	const { saveEntityRecord } = useDispatch( 'core' );
+	const [ product, setProduct ] = useState< Product | undefined >(
+		undefined
+	);
+
+	useEffect( () => {
+		saveEntityRecord( 'postType', 'product', {
+			title: AUTO_DRAFT_NAME,
+			status: 'auto-draft',
+			// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+			// @ts-ignore Incorrect types.
+		} ).then( ( autoDraftProduct: Product ) => {
+			setProduct( autoDraftProduct );
+		} );
+	}, [] );
+
+	return product;
+};
+
+const useProductEntity = ( productId: string ) => {
+	const product = useSelect( ( select: typeof WPSelect ) => {
+		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+		// @ts-ignore Missing types.
+		const { getEditedEntityRecord } = select( 'core' );
+
+		return getEditedEntityRecord( 'postType', 'product', productId );
+	} );
+
+	return product;
+};
+
+export function useProductEntityRecord(
+	productId: string | undefined
+): Product | undefined {
+	const selectedHook = useMemo( () => {
```

Comment:
Normally we don't want to allow the possibility of different hooks being fired in the same component on re-render, and this shouldn't allow that as only one would ever be fired on a particular screen. That said, if it feels uncomfortable we could just stick with something like the original approach using conditional components.

Reply:
I end up updating this, as React does have a specific rule to not call hooks conditionally: https://legacy.reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level

This does work and is a nifty a trick around it, but I don't want to run into any edge cases.